docs/design/NsfsVersioning.md

```diff
@@ -0,0 +1,92 @@
+        * Delete object/Delete multiple object of latest version/a given version
+        * Get object/Head object of latest version/a given version
+    
+
+## SOLUTION
+The Versioning implementation consists of lazy creation of a hidden sub directory that will contain all versions of all objects under the parent directory except of the latest version. 
+The latest version will be located in the parent directory itself.
+
+
+### PROPOSED SOLUTION DIAGRAM
+The following diagram illustrates a directory tree of a versioning enabled bucket
+
+<div id="top" />
+<img src="/docs/design/images/nsfs_versioning_dir_tree.png" />
+In the figure, each "regular" directory contains a hidden .versions/ sub directory that stores past versions of objects located under the parent directory, while the latest version of each object can be found under the parent directory itself. For instance, the latest version of Obj1 can be found under the root directory (/), while the old versions of Obj1 reside under /.versions/.
+
+
+## CONCEPTS 
+
+* Objects that are stored in a bucket before you set the versioning state have a version ID of null. 
+* Objects that are stored in a bucket After you set the versioning state have a unique version ID. 
+* PUT, POST and COPY operations will create a new version of the object identified by a unique version id.
+* DELETE latest version will create a delete marker which is a dummy version identified by a  unique version id as well.
+* DELETE version id will delete the version completely.
+
+
+### Version ID 
+* Will be allocated to a version on PUT/POST request (or DELETE version id). 
+* Consisted of the version mtimeMS + inode number.
+* The version ID will be attached as an xattr of the file.
+* When latest versions is moved to past versions directory, the file name will be changed from key to key_{version_id}.
+
+
+### Delete marker
+* A dummy file that will be created under the hidden .versions/ directory on DELETE latest request.
+* When A delete marker is the latest version of an object, it indicates that the object is deleted.
+* A unique version id will be allocated to the delete marker similiarly tor regular versions.
+  
+### Posix safe rename
+
+#### In the following cases NooBaa will move files between a directory and its .versions/ directory:
+
+* Put object / Complete multipart upload - a new latest version will be created and the old latest should move to .versions/.
+* Delete latest version - The latest version should move from parent directory to .versions/ & a delete marker will be created in .versions as well.
+* Delete version id (latest version & regular version) - The current latest version should be removed and the second latest should move from .versions/ to the parent directory.
+* Delete version id (latest version & a delete marker) - The current latest version should be removed and if the second latest is not a delete marker it should move from .versions/ to the parent directory.
```

Comment:
I think those cases are harder to understand, maybe an illustrative diagram (or a link) will be helpful.
Does the last case ('Delete version id (latest version & a delete marker)') is recursive? If yes, maybe it is worth mentioning it.

Reply:
I added a link to S3 versioning doc, in there, you will find that behind-the-scenes versions of the same object are stacked. I think that this idea is important in order to understand these scenarios.
Also, illustrations can be found in the detailed downstream documentation.
The last case is just a deletion of a version that is the latest version but it's also a delete marker